src/man/grepmidi.1
.\" grepmidi -- display structure of MIDI files
.\" Copyright (c) 2006 Daniel Kobras
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
.\"
.\" This manual page was written especially for Debian GNU/Linux.
.\"
.TH GREPMIDI 1 "July 2006"
.SH NAME
grepmidi \- display structure of MIDI files
.SH SYNOPSIS
\fBgrepmidi\fR [ \fI-f\fR ] \fIfile\fR [ \fI...\fR ]
.SH DESCRIPTION
The \fBgrepmidi\fR utility provides a simple parser for files in raw MIDI
format.  It is far from complete, but allows to get a grip of the basic
song structure in \fIfile\fR.  That is, it does not display single note,
but only meta information embedded in the MIDI file.  \fBgrepmidi\fR can
also be useful to check whether a given MIDI file contains errors.
.PP
More than one \fIfile\fR can be given on the commandline.  Option \fB-f\fR
tells \fBgrepmidi\fR to prepend each line of output with the filename that
is currently checked.
.PP
\fBgrepmidi\fR is distributed along with \fBMusE\fR, a full-fledged MIDI
Music Editor.
.SH "SEE ALSO"
.BR muse4 (1).
.SH AUTHORS
.B grepmidi
was written by Werner Schweer <ws@seh.de>.
.PP
This manual page was written by Daniel Kobras <kobras@debian.org> for the
Debian GNU/Linux system (but may be used by others).


ChangeLog
06.01.2014:
         - Midi instrument can now be changed from midi track info panel. (Tim) 
         - Some minor code tweaks, code cleanup in mtrackinfo.cpp, confmport.cpp. 
06.01.2014:
         - Added (and fixed) Roland SD-50 .idf file by Patrick. Needs more work: split in two .idfs (Tim)
         - Added GM2 .idf file created by me Tim.
         - Fixed a couple of improper "&apos" xml substitutions in xml.cpp
17.10.2013:
         - Fixed drag&drop plugins sometimes not working when built with optimize (rj)
10.10.2013:
         - Added a visual metronome to Bigtime (rj)
09.10.2013:
         - Changed track dragging to move instead of swap (rj)
         - Removed projects from 'recent' menu if they no longer exist (rj)
08.10.2013:
         - Added line drawing of tempo in graphical master track editor (rj)
         - Fixed bug (issue #342 Moving Aux sends crashes muse) (rj)
03.10.2013:
         - Added keyboard shortcuts in midi editors for functions, Step record, Midi input,
           Play events, Inc/Dec velocity (rj)
28.09.2013:
         - Changed default shortcut in PianoRoll for moving to current position from C to I
           the former was conflicting with the default global shortcut for metronome (rj)
27.09.2013:
         - Changed beat lines to a darker shade (rj)
02.09.2013:
         - Display git hash and date in the about box instead of svn revision (Orcan)
         - Bumped the year string to 2013 in the about box
01.09.2013:
         - Mods/fixes to Midi Input Transformator plugin: (Tim)
           Fixed some bugs when reading procVal2Op from song.
           'Program' is now allowed 'Event type' selection.
           'Toggle' is a new Processing Value#2 Operation. It toggles between Value#1 and Value#2.
           This means for example a standard midi foot controller (which only sends program changes)
            can now be used to send TOGGLED midi CC values, say back to an effects unit
            thus the foot controller becomes like regular 'chained' stomp boxes, operating
            upon individual effects.
14.05.2013:
         - When midi in is enabled in drum editor the selected track is moved along with
           the triggered key (rj)
12.05.2013:
         - Fixed bug with playback of drums clicking on notes in the new drum editor,
           it was playing the wrong instrument (rj)
08.05.2013:
         - Yet another MAJOR audio engine and plugin/synth process chain re-write. (Tim...)
         - Track controllers (vol, pan) now sample-accurate.
         - Track controllers (vol, pan) slope limited for near-noiseless operation. TODO: User settings.
         - Fixed: "Duplicate Selected Tracks" crashes. Also now copies plugins and synths too.
         - DSSI + Native VST: Fixed not remembering generic/native GUI pos/size. TODO: DSSI native GUI.
         - Fixed regression by flo: Sliders were recording multiple automation points at same value.
         - Fixed canvases not remembering position/zoom - everything reopens at exact same pos/zoom now.
         - Fixed automatable audio GUI controls 'jumpy' in TOUCH mode if heavy graphics causing slowdowns.
           When pressed, any control now truly 'disengages' from competing/intefering controller stream.
         - Improved audio automation modes, + READ mode can now be interrupted by GUI or ext control.
         - MESS synths (esp Deicsonze): Controls (like Track Info program) and synth controls now mirror
            each other, both ways.
         - Deicsonze softsynth: Fixed several problems: Not remembering settings + ladspa plugin settings,
            midi controllers were in wrong domain - moved to NRPN14 type. TODO: Fix 'biased' midi controllers
            like pan and detune etc.
         - Native VST: Call idle periodically. Makes some plugins like Glitch work, as per LAD mail.
05.04.2013:
         - Change window title when there are unsaved changes (rj)
         - Add auto-save feature, when enabled tries to save after 5 minutes
           when there has been changes, unless transport is running (rj)
04.04.2013:
         - Added pitch control to SimpleDrums, first version (rj)
         - Fixed Ctrl+arrows in the arranger so the track list is scrolled (rj)
29.03.2013:
         - MusE 2.1.2 released! (rj)
28.03.2013:
         - Allow reading old drummaps for new style drumtracks (rj)
         - Added metronome icon in main window (rj)
27.03.2013:
         - Fixed moving events with keyboard in Drum editor (rj)
10.03.2013:
         - Added color settings for: midi events in arranger,wave events in arranger, wave events in wave editor (rj)
05.03.2013:
         - Added shortcut printout of tools in tool menu (rj)
03.02.2013:
         - Generalized theming somewhat more and added a basic Ardour complementary theme (rj)
28.02.2013:
         - Wave Editor: Fixed some painting problems. Optimizations, speed boosts. (Tim)
24.02.2013:
         - Improved dark theme, midi editors now set background color, the current part of the ruler is darker.
           Theme setting separated from the regular Ok button in the Appearance settings. Also fixed the color 
           viewer in the color selector when using the dark theme. (rj)
23.02.2013:
         - Added missing line draw shortcut (F) to drum editor. (Tim)
22.02.2013:
         * New: Install pre-built PDF + single/split HMTL docs. Separate devel docs. Added build script. (Tim)
           Build script is /doc/build_docs.sh You'll want to use it or a variant to get the HTML .aux files.
           Fixed table in my Appendix in tex file so it works with HTML.
         - For now, I've changed F1 help to open PDF, if not found fallback to HTML.
           I've #define MUSE_USE_PDF_HELP_FILE in help.cpp. Undefine for HTML only.
           May at least need to keep track of viewer process (don't open copies), also try context-sensitive help.
           Help opener always been LANG ready (looks for _xx) but this new build script is not 'automatic'.

src/grepmidi/grepmidi.cpp
//=========================================================
//  MusE
//  Linux Music Editor
//  $Id: grepmidi.cpp,v 1.1.1.1.2.1 2009/03/09 02:05:17 terminator356 Exp $
//
//  (C) Copyright 1999/2000 Werner Schweer (ws@seh.de)
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; version 2 of
//  the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
//=========================================================

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

#define XCHG_SHORT(x) ((((x)&0xFF)<<8) | (((x)>>8)&0xFF))
#ifdef __i486__
#define XCHG_LONG(x) \
     ({ int __value; \
        asm ("bswap %1; movl %1,%0" : "=g" (__value) : "r" (x)); \
       __value; })
#else
#define XCHG_LONG(x) ((((x)&0xFF)<<24) | \
		      (((x)&0xFF00)<<8) | \
		      (((x)&0xFF0000)>>8) | \
		      (((x)>>24)&0xFF))
#endif

#if __BYTE_ORDER == __LITTLE_ENDIAN
#define BE_SHORT(x) XCHG_SHORT(x)
#define BE_LONG(x) XCHG_LONG(x)
#else
#define BE_SHORT(x) x
#define BE_LONG(x) x
#endif


int cpos;
bool printName = false;
const char* curName;

//---------------------------------------------------------
//   readLong
//---------------------------------------------------------

int readLong(FILE* f)
      {
      int format;
      if (fread(&format, 4, 1, f) != 1) {
            printf("read long failed\n");
            exit(1);
            }
      cpos += 4;
      return BE_LONG(format);
      }

//---------------------------------------------------------
//   readShort
//---------------------------------------------------------

int readShort(FILE* f)
      {
      short format;
      if (fread(&format, 2, 1, f) != 1) {
            printf("read short failed\n");
            exit(1);
            }
      cpos += 2;
      return BE_SHORT(format);
      }

/*---------------------------------------------------------
 *    getvl
 *    Read variable-length number (7 bits per byte, MSB first)
 *---------------------------------------------------------*/

int getvl(FILE* f)
      {
      int l = 0;
      for (int i = 0;i < 8; i++) {
            int c = getc(f);
            ++cpos;
            if (c == EOF)
                  break;
            c &= 0xff;
            l += (c & 0x7f);
            if (!(c & 0x80))
                  return l;
            l <<= 7;
            }
      printf("Variable Len too long\n");
      return -1;
      }

//---------------------------------------------------------
//   skip
//---------------------------------------------------------

void skip(FILE* f, int n)
      {
      while (n--) {
            ++cpos;
            if (getc(f) == EOF) {
                  printf("skip %d failed\n", n);
                  exit(1);
                  }
            }
      }

//---------------------------------------------------------
//   grepTrack
//---------------------------------------------------------

int grepTrack(FILE* f, int trackno)
      {
//      printf("TRACK %d\n", trackno);
      int mtype, mlen;
      char* buffer;

      char tmp[4];
      fread(tmp, 4, 1, f);
      if (memcmp(tmp, "MTrk", 4))
            return -5;
      int len = readLong(f);
      int endpos = cpos + len;
      int runstate = -1;

      for (;;) {
/*            int nclick = */ getvl(f);
            int me     = getc(f);
            ++cpos;
            int a = -1;
            if ((me & 0x80) == 0) {
                  if (runstate == -1)
                        return -6;
                  a  = me;
                  me = runstate;
                  }
            switch (me & 0xf0) {
                  case 0x80:
                  case 0x90:
                  case 0xa0:
                  case 0xb0:
                  case 0xe0:
                        getc(f);
                        ++cpos;
                  // NOTE: Error suppressor for new gcc 7 'fallthrough' level 3 and 4:
                  // FALLTHROUGH
                  case 0xc0:
                  case 0xd0:
                        if (a == -1) {
                              a = getc(f);
                              ++cpos;
                              }
                        runstate = me;
                        break;
                  case 0xf0:
                        switch(me & 0xf) {
                              case 0:   // SYSEX
                                    mlen = getvl(f);
                                    skip(f, mlen);
                                    break;
                              case 1:
                              case 2:
                              case 3:
                              case 4:
                              case 5:
                              case 6:
                                    if (printName)
                                          printf("%s: ", curName);
                                    printf("unknown Message\n");
                                    break;
                              case 7:     // EOX
                              case 8:     // timing clock
                              case 9:     // undefined
                              case 0xa:   // start
                              case 0xb:   // continue
                              case 0xc:   // stop
                              case 0xd:   // undefined
                              case 0xe:   // active sensing
                                    if (printName)
                                          printf("%s: ", curName);
                                    printf("RT Message??\n");
                                    break;
                              case 0xf:   // META
                                    mtype = getc(f);
                                    ++cpos;
                                    mlen  = getvl(f);
                                    buffer = new char[mlen+1];
                                    if (mlen) {
                                          if (fread(buffer, mlen, 1, f) != 1) {
                                                if (printName)
                                                      printf("%s: ", curName);
                                                printf("---meta %d too short (%d)\n", mtype, mlen);
                                                exit(1);
                                                }
                                          cpos += mlen;
                                          }
                                    switch(mtype) {
                                          case 0x2f:
                                                delete[] buffer;
                                                goto end;
                                          case 1:
                                          case 2:
                                          case 3:
                                          case 4:
                                          case 5:
                                          case 6:  // Marker
                                          case 7:
                                          case 8:
                                          case 9:
                                          case 10:
                                          case 11:
                                          case 12:
                                          case 13:
                                          case 14:
                                          case 15:
                                                buffer[mlen] = 0;
                                                if (printName)
                                                      printf("%s: ", curName);
                                                printf("%02d Meta %0d: <%s>\n", trackno, mtype, buffer);
                                          default:
                                                break;
                                          }
                                    delete[] buffer;
                                    break;
                              }
                        break;
                  }
            }
end:
      if (cpos != endpos) {
            if (printName)
                  printf("%s: ", curName);
            printf("   %d zu kurz\n", cpos - endpos);
            }
      if (endpos-cpos > 0)
            skip(f, endpos-cpos);
      return 0;
      }

//---------------------------------------------------------
//   grep
//---------------------------------------------------------

int grep(FILE* f)
      {
      char tmp[4];

      fread(tmp, 4, 1, f);
      if (memcmp(tmp, "MThd", 4) != 0)
            return -2;
      int len = readLong(f);
      if (len < 6)
            return -3;
      cpos += 8;
      int format     = readShort(f);
      int ntracks    = readShort(f);
      readShort(f);     // division
      if (len > 6)
            skip(f, len-6); /* skip the excess */
      int rv = 0;
      switch (format) {
            case 0: rv = grepTrack(f, 0); break;
            case 1:
                  for (int i = 0; i < ntracks; i++) {
                        rv = grepTrack(f, i);
                        if (rv != 0)
                              return rv;
                        }
                  break;
            default:
                  return -4;
            }

      return 0;
      }

//---------------------------------------------------------
//   grep
//---------------------------------------------------------

int grepMidi(const char* name)
      {
      curName = name;
      const char*p = strrchr(name, '.');
      FILE* f;
      if (p && strcmp(p, ".gz") == 0) {
            char buffer[512];
            snprintf(buffer, 512, "gunzip < %s", name);
            f = popen(buffer, "r");
            }
      else {
            p = 0;
            f = fopen(name, "r");
            }
      if (f == 0)
            return -1;
      cpos = 0;
      int rc = grep(f);
      if (p)
            pclose(f);
      else
            fclose(f);
      return rc;
      }

//---------------------------------------------------------
//   usage
//---------------------------------------------------------

void usage(const char* fname, const char* txt)
      {
      fprintf(stderr, "%s: %s\n", fname, txt);
//      fprintf(stderr, "usage:\n");
      }

//---------------------------------------------------------
//   main
//---------------------------------------------------------

int main(int argc, char* argv[])
      {
      int c;
      while ((c = getopt(argc, argv, "f")) != EOF) {
            switch (c) {
                  case 'f': printName = true; break;
                  default:  usage(argv[0], "bad argument"); return -1;
                  }
            }
      argc -= optind;
      ++argc;
      const char* p = 0;
      for (int i = 1; i < argc; ++i) {
            switch (grepMidi(argv[i])) {
                  case 0:     break;
                  case -1:    p = "not found"; break;
                  case -2:    p = "no 'MThd': not a midi file"; break;
                  case -3:    p = "file too short"; break;
                  case -4:    p = "bad file type"; break;
                  case -5:    p = "no 'MTrk': not a midi file"; break;
                  case -6:    p = "no running state"; break;
                  default:
                        printf("was??\n");
                        return -1;
                  }
            }
      if (p)
            printf("Error: <%s>\n", p);
      return 0;
      }
